### PR TITLE
Typed events for db:add-missing-…

### PR DESCRIPTION
**Pull request by nickvergessen**
* One step closer to #38546 

## Summary

* Instead of the last commit [fix(db)!: Remove private legacy event because we can not keep it](https://github.com/nextcloud/server/commit/d83bb944915d26872163154b68b749391ebc56c0) We could also add a "dispatchLegacy" to `IEventDispatcher` that allows to dispatch symfony's `GenericEvent`. However that one is also deprecated and will be dropped sooner or later. From my POV introducing yet another legacy layer is just too much (especially since I couldn't find any listeners), and dropping is the better option.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Comment by kesselb** [comment]
Unrelated, but interesting: If you put your instance into maintenance mode to add missing indexes, apps cannot register their indexes because the app and event listener are not loaded, and the index is not added.[^1]

[^1]: https://github.com/nextcloud/server/pull/37178

File: core/Command/Db/AddMissingIndices.php [hunk 1]
```diff
@@ -91,7 +88,22 @@ protected function execute(InputInterface $input, OutputInterface $output): int
 					$table = $schema->getTable($missingIndex['tableName']);
 					if (!$table->hasIndex($missingIndex['indexName'])) {
 						$output->writeln('<info>Adding additional ' . $missingIndex['indexName'] . ' index to the ' . $table->getName() . ' table, this can take some time...</info>');
-						$table->addIndex($missingIndex['columns'], $missingIndex['indexName']);
+
+						if ($missingIndex['dropUnnamedIndex'])
+						foreach ($table->getIndexes() as $index) {
```
Comment (come-nc) [inline core/Command/Db/AddMissingIndices.php:93]: Misleading indentation here, see php-cs failure.

File: core/Application.php [hunk 2]
```diff
@@ -88,244 +84,237 @@ public function __construct() {
...
-					$table = $schema->getTable('comments');
-
-					if (!$table->hasColumn('reference_id')) {
-						$subject->addHintForMissingColumn($table->getName(), 'reference_id');
-					}
-				}
-			}
-		);
+		$eventDispatcher->addListener(AddMissingIndicesEvent::class, function (AddMissingIndicesEvent $event) {
+			$event->addMissingIndex(
```
Comment (juliushaertl) [inline core/Application.php:88]: Maybe move this out to a dedicated listener class for core?

File: core/Command/Db/AddMissingIndices.php (hunk 1 above)
Comment (nickvergessen) [reply core/Command/Db/AddMissingIndices.php]: Fixed

File: core/Application.php (hunk 2 above)
Comment (nickvergessen) [reply core/Application.php:88]: Can we do that in a follow up? Just trying to get closer to #38546  ^_^
cleaning up the core application is something for another day

Comment (juliushaertl) [reply core/Application.php:88]: Sure

Comment (ChristophWurst) [reply core/Application.php:88]: > for another day

Today?

**Review by ChristophWurst — APPROVED** [review]

**Review by miaulalala — APPROVED** [review]

**Review by bigcat88 — APPROVED** [review]

**Comment by nickvergessen** [comment]
Docs in https://github.com/nextcloud/documentation/pull/10858

**Pull request merged**